### PR TITLE
feat: add filter lambda udf

### PR DESCRIPTION
**Pull request by lct45**
### Description 
Adds `filter()` udf/lambda invocation function so users can filter down arrays + maps

### Testing done 
Unit test + qtt

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/spec.json [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+          } ],
+          "default" : null
+        } ],
+        "connect.name" : "io.confluent.ksql.avro_schemas.KsqlDataSourceSchema"
+      },
+      "valueFormat" : "AVRO",
+      "replicas" : 1,
+      "numPartitions" : 4
+    } ],
+    "statements" : [ "CREATE STREAM test (ID STRING KEY, map MAP<STRING, INTEGER>) WITH (kafka_topic='test_topic', value_format='AVRO');", "CREATE STREAM OUTPUT AS SELECT ID, filter(map, (x,y) => (len(x) * 5 > 15) AND y % 5 = 0) AS lambda FROM test;" ],
```
Comment (guozhangwang) [inline ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/spec.json:144]: Wondering why use avro for maps and json for lists?

File: ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/plan.json [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        },
+        "valueFormat" : {
+          "format" : "AVRO"
+        }
+      },
+      "orReplace" : false
+    }
+  }, {
+    "@type" : "ksqlPlanV1",
+    "statementText" : "CREATE STREAM OUTPUT AS SELECT\n  TEST.ID ID,\n  FILTER(TEST.MAP, (X, Y) => (((LEN(X) * 5) > 15) AND ((Y % 5) = 0))) LAMBDA\nFROM TEST TEST\nEMIT CHANGES",
```
Comment (guozhangwang) [inline ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/plan.json:22]: For my education: is the most-outer brackets required by the syntax? I.e. is 

```
(X, Y) => (LEN(X) * 5) > 15 AND (Y % 5) = 0 LAMBDA
```

Acceptable? If not, is

```
(X, Y) => ((LEN(X) * 5) > 15) AND ((Y % 5) = 0) LAMBDA
```

Acceptable?

File: ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_filter_a_mixed_map/7.0.0_1614034662398/plan.json [hunk 3]
```diff
@@ -0,0 +1,145 @@
+{
+  "plan" : [ {
+    "@type" : "ksqlPlanV1",
+    "statementText" : "CREATE STREAM TEST (ID BIGINT KEY, VALUE MAP<STRING, ARRAY<INTEGER>>) WITH (KAFKA_TOPIC='test_topic', KEY_FORMAT='KAFKA', VALUE_FORMAT='AVRO');",
```
Comment (guozhangwang) [inline ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_filter_a_mixed_map/7.0.0_1614034662398/plan.json:4]: nit: the file name `mixed_map` is a bit misleading, how about `..nested_map`?

File: ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/plan.json (hunk 2 above)
Comment (lct45) [reply ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/plan.json:22]: The outermost brackets are added by the test generator/parser (not sure which). The initial syntax for the test was `filter(map, (x,y) => (len(x) * 5 > 15) AND y % 5 = 0) AS lambda` and I just confirmed that it also works with `filter(map, (x,y) => len(x) * 5 > 15 AND y % 5 = 0) AS lambda` and `filter(map, (x,y) => ((len(x) * 5) > 15) AND y % 5 = 0) AS lambda`. Does that answer your question?

File: ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/spec.json (hunk 1 above)
Comment (lct45) [reply ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/spec.json:144]: Ah, that was just random. I think there's JSON for 1 list and then everything else is AVRO. Do you think it's better to have them more distributed?

File: ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/plan.json (hunk 2 above)
Comment (guozhangwang) [reply ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/plan.json]: Yup! That's good enough for me.

File: ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/spec.json (hunk 1 above)
Comment (guozhangwang) [reply ksqldb-functional-tests/src/test/resources/historical_plans/filter_-_apply_filter_lambda_function_to_int_map/7.0.0_1614088997381/spec.json]: I think it's fine, just wondering if it is intentional for reasons that I was not aware :)

**Review by guozhangwang — APPROVED** [review]

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/lambda/Filter.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    if (array == null) {
+      return null;
+    }
+    return array.stream().filter(function::apply).collect(Collectors.toList());
+  }
+
+  @Udf(description = "When filtering a map, "
+      + "the function provided must have a boolean result. "
+      + "For each map entry, the function will be applied to the "
+      + "key and value. The filtered map is returned."
```
Comment (stevenpyzhang) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/lambda/Filter.java:60]: can we specify what arguments are in what order for the function (key then value) you can look at https://github.com/confluentinc/ksql/pull/6994/files for example for reduce and transform of map

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/filter.json [hunk 5]
```diff
@@ -0,0 +1,100 @@
+{
+  "comments": [
+    "Tests covering the filter lambda."
+  ],
+  "tests": [
```
Comment (stevenpyzhang) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/filter.json:5]: Can we have all the test records have different keys? Makes things easier to keep track of which output record belongs to which input record when debugging

**Review by stevenpyzhang — APPROVED** [review]
lgtm with minor comments

**Pull request merged**